### ISSUE TITLE
HTTPS request stucked with proxy

### ISSUE THREAD

**Issue by kotori2** [yhirose/cpp-httplib#635]
Just some simple code like this. It was stucked on `cli.Get()` when I'm trying to request any site with HTTPS plus proxy.
This sample works perfectly with `httplib::Client` and my proxy supports HTTPS as well. Also I think I can't reproduce this on Windows. 
```
    httplib::SSLClient cli("ip.sb");
    cli.set_proxy("127.0.0.1", 6152);
    auto res = cli.Get("/", {
            {"User-Agent", "curl/7.64.1"}
    });
    std::cout << res.error() << std::endl;
    std::cout << res->body << std::endl;
```
Some logs showing that my proxy is working:
```
% curl https://ip.sb/ -v
* Uses proxy env variable https_proxy == 'http://127.0.0.1:6152'
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 6152 (#0)
* allocate connect buffer!
* Establish HTTP proxy tunnel to ip.sb:443
> CONNECT ip.sb:443 HTTP/1.1
> Host: ip.sb:443
> User-Agent: curl/7.64.1
> Proxy-Connection: Keep-Alive
>
< HTTP/1.0 200 Connection established
<
* Proxy replied 200 to CONNECT request
* CONNECT phase completed!
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* CONNECT phase completed!
* CONNECT phase completed!
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-CHACHA20-POLY1305
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=ip.sb
*  start date: Aug 20 09:25:58 2020 GMT
*  expire date: Nov 18 09:25:58 2020 GMT
*  subjectAltName: host "ip.sb" matched cert's "ip.sb"
*  issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7fbbf980e800)
> GET / HTTP/2
> Host: ip.sb
> User-Agent: curl/7.64.1
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 403
< server: nginx
< date: Wed, 02 Sep 2020 14:38:43 GMT
< content-type: text/html
< content-length: 146
< vary: Accept-Encoding
<
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
<hr><center>nginx</center>
</body>
</html>
* Connection #0 to host 127.0.0.1 left intact
* Closing connection 0
```

**Comment by yhirose** [comment]
@kotori2, thanks for the report. I actually work on macOS mainly, and I have just tested with `test/test_proxy.cc` and confirmed all the unit tests passed. (It tests against squid proxy server on Ubuntu on Docker.)

Run the test squid server on Ubuntu on Docker:
```bash
cd test/proxy
./up.sh
```

Run proxy unit test:
```bash
cd test
make test_proxy && ./test_proxy
``` 

Could you try the above on your macOS? Thanks for your help!

**Comment by kotori2** [comment]
It sounds like everything is working except `DigestAuthTest`s. But I think that's not related to my issue because my proxy doesn't use any kind of authentication, which was not covered in your tests. I'm going to try modifying squid config files and do some more tests.
```
make test_proxy && ./test_proxy
c++ -o test_proxy -ggdb -O0 -std=c++11 -DGTEST_USE_OWN_TR1_TUPLE -I.. -I. -Wall -Wextra -Wtype-limits -Wconversion test_proxy.cc gtest/gtest-all.cc gtest/gtest_main.cc -DCPPHTTPLIB_OPENSSL_SUPPORT -I/usr/local/opt/openssl@1.1/include -L/usr/local/opt/openssl@1.1/lib -lssl -lcrypto -DCPPHTTPLIB_ZLIB_SUPPORT -lz -DCPPHTTPLIB_BROTLI_SUPPORT -I/usr/local/opt/brotli/include -L/usr/local/opt/brotli/lib -lbrotlicommon -lbrotlienc -lbrotlidec -pthread
Running main() from gtest_main.cc
[==========] Running 12 tests from 4 test cases.
[----------] Global test environment set-up.
[----------] 4 tests from ProxyTest
[ RUN      ] ProxyTest.NoSSLBasic
[       OK ] ProxyTest.NoSSLBasic (8 ms)
[ RUN      ] ProxyTest.SSLBasic
[       OK ] ProxyTest.SSLBasic (6 ms)
[ RUN      ] ProxyTest.NoSSLDigest
[       OK ] ProxyTest.NoSSLDigest (6 ms)
[ RUN      ] ProxyTest.SSLDigest
[       OK ] ProxyTest.SSLDigest (5 ms)
[----------] 4 tests from ProxyTest (25 ms total)

[----------] 4 tests from RedirectTest
[ RUN      ] RedirectTest.YouTubeNoSSLBasic
[       OK ] RedirectTest.YouTubeNoSSLBasic (434 ms)
[ RUN      ] RedirectTest.YouTubeNoSSLDigest
[       OK ] RedirectTest.YouTubeNoSSLDigest (218 ms)
[ RUN      ] RedirectTest.YouTubeSSLBasic
[       OK ] RedirectTest.YouTubeSSLBasic (515 ms)
[ RUN      ] RedirectTest.YouTubeSSLDigest
[       OK ] RedirectTest.YouTubeSSLDigest (595 ms)
[----------] 4 tests from RedirectTest (1762 ms total)

[----------] 2 tests from BaseAuthTest
[ RUN      ] BaseAuthTest.NoSSL
[       OK ] BaseAuthTest.NoSSL (2010 ms)
[ RUN      ] BaseAuthTest.SSL
[       OK ] BaseAuthTest.SSL (5058 ms)
[----------] 2 tests from BaseAuthTest (7068 ms total)

[----------] 2 tests from DigestAuthTest
[ RUN      ] DigestAuthTest.SSL
test_proxy.cc:185: Failure
Value of: res != nullptr
  Actual: false
Expected: true
[  FAILED  ] DigestAuthTest.SSL (2123 ms)
[ RUN      ] DigestAuthTest.NoSSL
test_proxy.cc:194: Failure
Value of: res->status
  Actual: 407
Expected: 401
test_proxy.cc:194: Failure
Value of: res->status
  Actual: 407
Expected: 401
test_proxy.cc:194: Failure
Value of: res->status
  Actual: 407
Expected: 401
test_proxy.cc:194: Failure
Value of: res->status
  Actual: 407
Expected: 401
[  FAILED  ] DigestAuthTest.NoSSL (3224 ms)
[----------] 2 tests from DigestAuthTest (5347 ms total)

[----------] Global test environment tear-down
[==========] 12 tests from 4 test cases ran. (14202 ms total)
[  PASSED  ] 10 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] DigestAuthTest.SSL
[  FAILED  ] DigestAuthTest.NoSSL

 2 FAILED TESTS
```

**Comment by kotori2** [comment]
Okay after some digging I found the issue actually came from this line:
https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L5037
The proxy I'm using is actually responding with a `HTTP/1.0 200 Connection established`, idk why you are closing connection after receiving this. Also it was stucked because deadlock occurred in `stop_core()`

**Comment by kotori2** [comment]
Some stack trace here
<img width="706" alt="image" src="https://user-images.githubusercontent.com/17663689/92031037-84fad900-ed9a-11ea-8dfb-eb008bf614a2.png">

**Comment by yhirose** [comment]
@kotori2, thanks for the find the problem. I have simply forgotten to handle the 'Connection established' extra header line massage, since I don't use proxies that produces the extra message. Before I make a change, could you comment out the following lines, and see if it fixes your proxy sever connection problem?

https://github.com/yhirose/cpp-httplib/blob/b0fd4befb11d8f02af777f23b2562cf6078a28c4/httplib.h#L5037-L5040

**Comment by kotori2** [comment]
Yes, I already tried deleting `res.version == "HTTP/1.0"` check and it works

**Comment by yhirose** [comment]
@kotori2, could you test with the version in #637?

**Comment by kotori2** [comment]
Got EXC_BAD_ACCESS at `res.get_header_value("Connection")`.
smh weird that why I didn't reproduced this before

**Comment by yhirose** [comment]
@kotori2, could you explain more about EXC_BAD_ACCESS, because it's technically impossible to happen according to the code itself?

```cpp
inline std::string Response::get_header_value(const char *key,
                                              size_t id) const {
  return detail::get_header_value(headers, key, id, "");
}
```

**Comment by kotori2** [comment]
I tried again with build cache cleaned and it won't reproduce, so I think it's working now. Thanks for your hard work!

**Comment by yhirose** [comment]
@kotori2, do you mean that you don't need the fix in #637 any more? I am get confused with your comments...

**Comment by yhirose** [comment]
@kotori2, thanks for the clarification. I have just merged #637. Thanks for your contribution!
